openoffload/go/server/sessionoffload.go

```diff
@@ -127,54 +168,72 @@ func next_session_id() (uint64, error) {
 
 func (s *server) AddSession(stream fw.SessionTable_AddSessionServer) error {
 	var total int32
+	var resp fw.AddSessionResponse
+	var add_new_session bool
 
 	for {
+		add_new_session = true
 		sr, err := stream.Recv()
 		if err == io.EOF {
-			return stream.SendAndClose(&fw.AddSessionResponse{
-				RequestStatus: fw.AddSessionStatus__SESSION_ACCEPTED,
-				ErrorStatus:   0,
-			})
+			return stream.SendAndClose(&resp)
 		}
 		if err != nil {
 			return err
 		}
 
+		// Check if this session already exists
+		if existing_session := find_session_by_7_tuple(sr.InLif, sr.OutLif,
+								sr.SourceIp, sr.SourceIpv6,
+								sr.DestinationIp, sr.DestinationIpv6,
+								sr.SourcePort, sr.DestinationPort,
+								&sr.ProtocolId, &sr.IpVersion); existing_session != 0 {
+			log.Printf("Existing session with 7-tuple found")
+			session_resp_err := fw.SessionResponseError{
+				SessionId: existing_session,
+				ErrorStatus: fw.RequestStatus_value["_REJECTED_SESSION_ALREADY_EXISTS"],
+			}
+			resp.ResponseError = append(resp.ResponseError, &session_resp_err)
+			add_new_session = false
+		}
+
 		newSessionId, err := next_session_id()
 		if err != nil {
 			log.Printf("Error getting new session ID: %v", err)
-			return err
+			add_new_session = false
```

Comment:
What's the exit strategy here?  It won't add the session, but then will loop back and try again.  In this case, what's the expected behavior of the stream.Recv above on the next loop?

Reply:
This is handling a stream request, so we iterate on each request, and for the ones which fail (and for which we can get a session ID), we return an error. However, in this case, the fact that `next_session_id()` returns zero means the session table is full. So we should also append a `SessionResponseError` indicating that here. I'll push a commit to address this now. Great find Stephen!